### PR TITLE
[host] linux: updated XCB implementation and added cursor functionality

### PR DESCRIPTION
**Pull request by VMFortress**
# Summary
- Added `.initOptions` and `.stop` to XCB capture
- Added cursor support to XCB capture via `pointerThread`
# Known Issues
- XCB still requires `app:throttleFPS` as it will grab frames regardless of change
- No implementation of capture timings

### CODE REVIEW

File: host/platform/Linux/capture/XCB/src/xcb.c [hunk 1]
```diff
@@ -151,6 +149,27 @@ static bool xcb_init(void)
   }
   DEBUG_INFO("Frame Data       : 0x%" PRIXPTR, (uintptr_t)this->data);
 
+  xcb_query_extension_cookie_t extension_cookie =
+			xcb_query_extension(this->xcb, strlen("XFIXES"), "XFIXES");
+  xcb_query_extension_reply_t * extension_reply =
+			xcb_query_extension_reply(this->xcb, extension_cookie, NULL);
+  if(extension_reply == NULL)
```
Comment (gnif) [inline host/platform/Linux/capture/XCB/src/xcb.c:156]: No need for `{}` on single line If statements, eg:
```
if (a)
  b;
```

Same for if/else

```
if (a)
{
  b;
  c;
}
else
  d;
```

File: host/platform/Linux/capture/XCB/src/xcb.c [hunk 2]
```diff
@@ -293,30 +308,65 @@ static int pointerThread(void * unused)
     uint32_t size;
     if (!this->getPointerBufferFn(&data, &size))
     {
-      DEBUG_WARN("failed to get a pointer buffer");
+      DEBUG_WARN("Failed to get a pointer buffer");
```
Comment (gnif) [inline host/platform/Linux/capture/XCB/src/xcb.c:311]: Cosmetic changes should be in a separate patch.

File: host/platform/Linux/capture/XCB/src/xcb.c [hunk 3]
```diff
@@ -293,30 +308,65 @@ static int pointerThread(void * unused)
...
+    }
+    else
+    {
+      pointer.positionUpdate = false;
+    }
+
+    if(!pointer.positionUpdate && !pointer.shapeUpdate)
+    {
+      free(curReply);
+      nsleep(1000);
```
Comment (gnif) [inline host/platform/Linux/capture/XCB/src/xcb.c:352]: any reason this is `nsleep` and not `usleep(1)`? Also this is a very short sleep, i'd suggest this is too fast.

File: host/platform/Linux/capture/XCB/src/xcb.c [hunk 4]
```diff
@@ -151,6 +149,27 @@ static bool xcb_init(void)
...
+	DEBUG_ERROR("Extension \"XFIXES\" isn't available");
+  }
+  free(extension_reply);
+
+  xcb_xfixes_query_version_cookie_t version_cookie =
+			xcb_xfixes_query_version(this->xcb, XCB_XFIXES_MAJOR_VERSION, XCB_XFIXES_MINOR_VERSION);
+  xcb_xfixes_query_version_reply_t * version_reply =
+			xcb_xfixes_query_version_reply(this->xcb, version_cookie, NULL);
+
+  if(version_reply == NULL)
```
Comment (gnif) [inline host/platform/Linux/capture/XCB/src/xcb.c:167]: `if (!version_reply)` please

**Review by gnif — CHANGES_REQUESTED** [review]
Going to hold off reviewing more until this patch set is cleaned up. Please fix the commit formatting, squash relevent commits together and remove reverts.

Example commit format:
```
[host] xcb: moved xcb_stop to correct location

Some more information on what this change is
```

Also and fix the merge by re-basing, we do not allow merges here.
Your goal is to produce a PR that is easy to review/follow without a miss mash of reverts/changes and clear reasoning for each commit.

File: host/platform/Linux/capture/XCB/src/xcb.c [hunk 5]
```diff
@@ -124,13 +149,52 @@ static bool xcb_init(void)
   }
   DEBUG_INFO("Frame Data       : 0x%" PRIXPTR, (uintptr_t)this->data);
 
+  xcb_query_extension_cookie_t extension_cookie =
+			xcb_query_extension(this->xcb, strlen("XFIXES"), "XFIXES");
+  xcb_query_extension_reply_t * extension_reply =
+			xcb_query_extension_reply(this->xcb, extension_cookie, NULL);
+  if(extension_reply == NULL)
+  {
+	DEBUG_ERROR("Extension \"XFIXES\" isn't available");
```
Comment (quantum5) [inline host/platform/Linux/capture/XCB/src/xcb.c:158]:
```suggestion
    DEBUG_ERROR("Extension \"XFIXES\" isn't available");
```

File: host/platform/Linux/capture/XCB/src/xcb.c [hunk 6]
```diff
@@ -124,13 +149,52 @@ static bool xcb_init(void)
   }
   DEBUG_INFO("Frame Data       : 0x%" PRIXPTR, (uintptr_t)this->data);
 
+  xcb_query_extension_cookie_t extension_cookie =
+			xcb_query_extension(this->xcb, strlen("XFIXES"), "XFIXES");
```
Comment (quantum5) [inline host/platform/Linux/capture/XCB/src/xcb.c:153]:
```suggestion
      xcb_query_extension(this->xcb, strlen("XFIXES"), "XFIXES");
```

File: host/platform/Linux/capture/XCB/src/xcb.c [hunk 7]
```diff
@@ -124,13 +149,52 @@ static bool xcb_init(void)
...
+  xcb_query_extension_reply_t * extension_reply =
+			xcb_query_extension_reply(this->xcb, extension_cookie, NULL);
```
Comment (quantum5) [inline host/platform/Linux/capture/XCB/src/xcb.c:155]: Don't use tabs.

```suggestion
      xcb_query_extension_reply(this->xcb, extension_cookie, NULL);
```

File: host/platform/Linux/capture/XCB/src/xcb.c [hunk 8]
```diff
@@ -124,13 +149,52 @@ static bool xcb_init(void)
...
+  if(extension_reply == NULL)
+  {
+	DEBUG_ERROR("Extension \"XFIXES\" isn't available");
+  }
+  free(extension_reply);
+
+  xcb_xfixes_query_version_cookie_t version_cookie =
+			xcb_xfixes_query_version(this->xcb, XCB_XFIXES_MAJOR_VERSION, XCB_XFIXES_MINOR_VERSION);
```
Comment (quantum5) [inline host/platform/Linux/capture/XCB/src/xcb.c:163]:
```suggestion
      
      xcb_xfixes_query_version(this->xcb, XCB_XFIXES_MAJOR_VERSION, XCB_XFIXES_MINOR_VERSION);
```

File: host/platform/Linux/capture/XCB/src/xcb.c [hunk 9]
```diff
@@ -124,13 +149,52 @@ static bool xcb_init(void)
...
+  xcb_xfixes_query_version_reply_t * version_reply =
+			xcb_xfixes_query_version_reply(this->xcb, version_cookie, NULL);
```
Comment (quantum5) [inline host/platform/Linux/capture/XCB/src/xcb.c:165]:
```suggestion
      xcb_xfixes_query_version_reply(this->xcb, version_cookie, NULL);
```

File: host/src/app.c [hunk 10]
```diff
@@ -756,7 +756,7 @@ int app_main(int argc, char * argv[])
       const uint64_t delta = now - previousFrameTime;
       if (delta < throttleUs)
       {
-        nsleep((throttleUs - delta) * 1000);
+        usleep(throttleUs - delta);
```
Comment (quantum5) [inline host/src/app.c:759]: `nsleep` should be preferred on code that executes on Windows since it is implemented with `NtDelayExecution` instead of whatever method MinGW uses.

File: host/platform/Linux/capture/XCB/src/xcb.c [hunk 11]
```diff
@@ -228,13 +293,94 @@ static CaptureResult xcb_getFrame(FrameBuffer * frame,
...
+
+    pointer.hx             = curReply->xhot;
+    pointer.hy             = curReply->yhot;
+    pointer.visible        = true;
+    pointer.x              = curReply->x;
+    pointer.y              = curReply->y;
+    pointer.format         = CAPTURE_FMT_COLOR;
+    pointer.width          = curReply->width;
+    pointer.height         = curReply->height;
+    pointer.pitch          = curReply->width * 4;
```
Comment (quantum5) [inline host/platform/Linux/capture/XCB/src/xcb.c:364]: Excessive indentation of `=`.

File: host/platform/Linux/capture/XCB/src/xcb.c [hunk 12]
```diff
@@ -228,13 +293,94 @@ static CaptureResult xcb_getFrame(FrameBuffer * frame,
...
+    }
+    else
+    {
+      pointer.positionUpdate = false;
+    }
+
+    if(!pointer.positionUpdate && !pointer.shapeUpdate)
+    {
+      free(curReply);
+      nsleep(1000);
```
Comment (quantum5) [inline host/platform/Linux/capture/XCB/src/xcb.c:352]: Why not get rid of the `continue` and move the `postPointerBufferFn` call into the `if` with logic toggled? This avoids duplicating the cleanup code.

File: host/src/app.c (hunk 10 above)
Comment (VMFortress) [reply host/src/app.c:759]: I only made this change as the original `nsleep((throttleUs - delta) * 1000);` did not appear to actually be sleeping for some reason (and the rest of `app.c` seemed to use `usleep`).

Any ideas as to why `nsleep` would not work properly in this file? The issue is reproducible yet `nsleep` seems to work fine in `xcb.c`.

File: host/platform/Linux/capture/XCB/src/xcb.c (hunk 3 above)
Comment (VMFortress) [reply host/platform/Linux/capture/XCB/src/xcb.c:352]: I did not have `#include <uinstd.h>` before so usleep was not included. I believe I originally intended it to be `usleep(1000)`/`sleep(1)` and scale back from there if the responsiveness wasn't optimal. I'll make that change.

The code did work without the sleep but obviously had much higher CPU usage than necessary, hence the addition. Ideally, it'd likely be better to sleep until a change but at the moment I wasn't sure how to do that in xcb given the poor documentation online.

File: host/src/app.c (hunk 10 above)
Comment (gnif) [reply host/src/app.c:759]: `nsleep` is part of our `common` lib, we have not really used/tested it under Linux to date, as quantum stated, it's a perf improvement for Windows. It might be bugged.

Comment (VMFortress) [reply host/src/app.c:759]: It looks like it was an issue with the types when using `1e9` in `nsleep`. I fixed that and reverted the sleep above to an `nsleep`.

File: common/include/common/time.h [hunk 13]
```diff
@@ -61,8 +61,8 @@ static inline void nsleep(uint64_t ns)
 #else
   const struct timespec ts =
   {
-    .tv_sec  = ns / 1e9,
-    .tv_nsec = ns - ((ns / 1e9) * 1e9)
+    .tv_sec  = ns / (uint64_t)1e9,
+    .tv_nsec = ns - ((ns / (uint64_t)1e9) * (uint64_t)1e9)
```
Comment (gnif) [inline common/include/common/time.h:65]: To inline with the other functions here this should be changed to
```c
.tv_sec  = (time_t)1000000000L,
.tv_nsec = ns - ((ns / 1000000000L) * 1000000000L)
```

**Review by gnif — APPROVED** [review]

**Comment by gnif** [comment]
Merged in 8ab130de
